### PR TITLE
LibRIFF+LibAudio+LibGfx: Refactor common (R)IFF handling into its own library

### PR DESCRIPTION
**Pull request by warpdesign**
@kleinesfilmroellchen first opened this [PR](21431) but is no longer motivated to update it so I'm taking over.

I only rebased the branch and fixed conflicts with recent changes made to `ILBMLoader`.

@Zaggy1024 You asked for changes but I'm not sure I understand what you mean:

>> From the commit message for the ISOBMFF changes:
>> RIFF ChunkID can (and should) be used as "just" a FourCC type, which
>> simplifies this code greatly.
>> 
> Could we make a helper/rename ChunkID to make this intended usage clearer?

### CODE REVIEW

**Comment by kleinesfilmroellchen** [comment]
> > Could we make a helper/rename ChunkID to make this intended usage clearer?

I'd like for a using definition for FourCC = ChunkID. Both names are used by two different specs (the former by ISOBMFF and IFF, the latter by RIFF) so I don't think it makes sense to prefer one over the other. A simple using definition will allow us to keep the implementation and functionality in sync

**Comment by Zaggy1024** [comment]
> I'd like for a using definition for FourCC = ChunkID. Both names are used by two different specs (the former by ISOBMFF and IFF, the latter by RIFF) so I don't think it makes sense to prefer one over the other. A simple using definition will allow us to keep the implementation and functionality in sync

I'd personally rather FourCC be the class name since that is more of a general term used by many formats, where ChunkID is a specific use case, but not really a big deal either way.

**Comment by warpdesign** [comment]
Do we all agree if I rename the `ChunkID` struct/file to `FourCC` or should I let it as is?

**Comment by trflynn89** [comment]
I'd like to propose an alternate template implementation:
https://github.com/trflynn89/serenity/commit/56870835d09e4f5be3d8414de510fc56f4f477ae

This makes the `RIFF::Detail` structures templated and moves the stream reading methods directly onto those classes. This lets us explicitly instantiate the entire class for RIFF and IFF in one line each (rather than one per function). And it lets us define the RIFF and IFF implementation with simple `using` aliases.

I personally find it easier to read/reason about, and it's a net-decrease in code, so just wanted to throw it out there. I haven't reviewed the rest of the PR.

**Comment by kleinesfilmroellchen** [comment]
> I'd like to propose an alternate template implementation: [trflynn89@5687083](https://github.com/trflynn89/serenity/commit/56870835d09e4f5be3d8414de510fc56f4f477ae)
> 
> This makes the `RIFF::Detail` structures templated and moves the stream reading methods directly onto those classes. This lets us explicitly instantiate the entire class for RIFF and IFF in one line each (rather than one per function). And it lets us define the RIFF and IFF implementation with simple `using` aliases.
> 
> I personally find it easier to read/reason about, and it's a net-decrease in code, so just wanted to throw it out there. I haven't reviewed the rest of the PR.

This is excellent, thanks for figuring out a way of templatizing the main classes without moving the implementations to the header (which was the thing I was struggling with). I will use your commit once I have time.

**Comment by warpdesign** [comment]
> I'd like to propose an alternate template implementation: [trflynn89@5687083](https://github.com/trflynn89/serenity/commit/56870835d09e4f5be3d8414de510fc56f4f477ae)
> 
> This makes the `RIFF::Detail` structures templated and moves the stream reading methods directly onto those classes. This lets us explicitly instantiate the entire class for RIFF and IFF in one line each (rather than one per function). And it lets us define the RIFF and IFF implementation with simple `using` aliases.
> 
> I personally find it easier to read/reason about, and it's a net-decrease in code, so just wanted to throw it out there. I haven't reviewed the rest of the PR.

@trflynn89 I'm not in position to discuss quite advanced C++ code like this, so if @kleinesfilmroellchen thinks it's better, I'm all for it! :) Do you want me to close this PR so that you can open one with your changes?

**Comment by trflynn89** [comment]
> Do you want me to close this PR so that you can open one with your changes?

Oh, definitely not :smile:  feel free to squash that commit into the "Rework to match LibGfx needs" commit.

**Comment by warpdesign** [comment]
> trflynn89@5687083

Ok! I'll see what I can do :)

**Comment by warpdesign** [comment]
@trflynn89 I squashed your changes: let me know if that looks good (at least ImageDecoder tests pass).

File: Meta/gn/secondary/Userland/Libraries/LibAudio/BUILD.gn [hunk 1]
```diff
@@ -13,7 +13,6 @@ shared_library("LibAudio") {
     "PlaybackStream.cpp",
     "QOALoader.cpp",
     "QOATypes.cpp",
-    "RIFFTypes.cpp",
```
Comment (ADKaster) [inline Meta/gn/secondary/Userland/Libraries/LibAudio/BUILD.gn:16]: FWIW if you wanted to update the gn build to work with these changes, you would need to add a new LibRIFF/BUILD.gn and add `//Userland/Libraries/LibRiff` as a dependency to LibAudio and LibGfx. But this smells like an IDE's "delete all references" sort of change :)

**Review by ADKaster — APPROVED** [review]
Let's move forward with this library refactor. One centralized location for parsing a particular data structure seems better than 3-4. In the future we should think about renaming the library to something more generic and putting more common classes in it, or moving the parsing of RIFF headers to such a generic library (people have mentioned 'LibMultimedia' as a desire in the past).

**Pull request merged**